Fix two scripted_metric bugs

nik9000 opened the pull request:
Fixes two bugs introduced by #57627:
1. We were not properly letting go of memory from the request breaker
   when the aggregation finished.
2. We no longer supported totally arbitrary stuff produced by the init
   script because we *assumed* that it'd be ok to run the script once
   and clone its results. Sadly, cloning can't clone *anything* that the
   init script can make, like `String` arrays. This runs the init script
   once for every new bucket so we don't need to clone.

nik9000 committed:
Fix two scripted_metric bugs

Fixes two bugs introduced by #57627:
1. We were not properly letting go of memory from the request breaker
   when the aggregation finished.
2. We no longer supported totally arbitrary stuff produced by the init
   script because we *assumed* that it'd be ok to run the script once
   and clone its results. Sadly, cloning can't clone *anything* that the
   init script can make, like `String` arrays. This runs the init script
   once for every new bucket so we don't need to clone.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

nik9000 commented:
I've marked this a `non-issue` because it fixes an unreleased bug.It was still a bug though!

jimczi commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregatorTests.java:
```diff
@@ -181,6 +200,46 @@ public static void initMockScripts() {
+    private CircuitBreakerService circuitBreakerService;
+
+    @Before
+    public void mockBreaker() {
```
We could check this for all aggs in the `AggregatorTestCase` directly ?

nik9000 replied:
Would you be ok if I moved it in there in a followup? I think there are quite a few fiddly things involved in getting that right.

jimczi replied:
Sure, it might uncovered some other bugs so be prepared ;).

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregator.java:
```diff
@@ -170,7 +180,7 @@ public InternalAggregation buildEmptyAggregation() {
 
     @Override
-    public void close() {
+    public void doClose() {
```
;)

nik9000 replied:
I'm dreaming of a checkstyle rule.....

jimczi reviewed (commented):
I left one question

jimczi reviewed (approved):
LGTM

nik9000 committed:
Merge branch 'master' into scripted_metric_mem_fixup

nik9000 committed:
Handle mutating params

Adds code to handle when the init script mutates the aggregation
parameters. I hadn't caught this in my initial testing and it is
*weird*, but we allow it.
